File: src/ast/node/type.ts
import type { Pass } from "../../pass/pass";
import type { Parameter } from "../tree/function/parameter";
import type { Block } from "./block";
import type { Lambda } from "./lambda";
import { Node } from "./node";

export class Type<T> extends Node<any> {
    public static Void = class extends Type<void> {}
    public static Any = class extends Type<any> {}
    public static String = class extends Type<string> {}
    public static Number = class extends Type<number> {}
    public static Boolean = class extends Type<boolean> {}
    public static Lambda = class extends Type<Lambda> {
        constructor(
            public parameters: Parameter[],
            public returnType: Type<any>,
            parent: any
        ) {
            super(parent);
        }
    }

    public isArray: boolean = false;

    public visit(pass: Pass): void {
        pass.visitType(this);
    }
}

File: src/main/kotlin/me/honkling/miniscript/parser/ast/Type.kt
package me.honkling.miniscript.parser.ast

import me.honkling.miniscript.parser.ast.expression.Expression
import me.honkling.miniscript.parser.ast.function.Parameter
import me.honkling.miniscript.pass.Pass
import me.honkling.miniscript.parser.ast.function.Function as MSFunction
import me.honkling.miniscript.parser.ast.prototype.Class as MSClass

abstract class Type<T>(parent: Node<*>? = null) : Node<Node<*>?>(null, parent) {
    object Char : Type<kotlin.Char>()
    object Number : Type<kotlin.Number>()
    object Boolean : Type<kotlin.Boolean>()
    class Array<T>(val innerType: Type<T>) : Type<kotlin.Array<T>>()
    object Dictionary : Type<HashMap<Any?, Any?>>()
    object Void : Type<Nothing?>()
    object Any : Type<Any>()
    class Class(val reference: Expression<*>, parent: Node<*>?) : Type<MSClass>(parent)
    class Function(
        val parameters: List<Parameter>,
        val returnType: Type<*>
    ) : Type<MSFunction>()

    override fun accept(pass: Pass) {
        pass.visitType(this)
    }
}